libs/generator/generator.cpp

```diff
@@ -0,0 +1,464 @@
+        }
+        for (const auto &elem : entity) {
+            const auto &key = elem.key_c_str();
+            auto expected_it = std::find_if(
+                expected_keys.begin(), expected_keys.end(),
+                [&elem](const char *expected_key) { return elem.key() == expected_key; });
+            if (expected_it == expected_keys.end()) {
+                error_msg << "Unexpected key '" << elem.key() << "' for " << result_type_name(res);
+                return false;
+            }
+            if (!key_handler(elem.key_c_str(), elem.value(), res)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    bool get_number(const std::string &key, const boost::json::value &json_value, size_t &result) {
+        if (json_value.is_uint64()) {
+            result = json_value.as_uint64();
+        } else if (json_value.is_int64()) {
+            result = json_value.as_int64();
+        } else {
+            error_msg << "Value '" << boost::json::serialize(json_value) << "' of '" << key
+                      << "' key is not a number";
+            return false;
+        }
+        return true;
+    }
+
+    // Transform hex string into Address or bytes
+    template<typename OutputContainer>
+    static bool get_bytes(const std::string &key, const boost::json::value hex_value,
+                          OutputContainer &result) {
+        if (!hex_value.is_string()) {
+            error_msg << "Value '" << boost::json::serialize(hex_value) << "' of '" << key
+                      << "' is not a string";
+            return false;
+        }
+        std::string hex_string = hex_value.as_string().c_str();
+        std::vector<uint8_t> bytes;
+        // Skip 0x prefix if exists
+        if (hex_string[0] == '0' && hex_string[1] == 'x') {
+            hex_string.erase(0, 2);
+        }
+        try {
+            boost::algorithm::unhex(hex_string, std::back_inserter(bytes));
+        } catch (const boost::algorithm::non_hex_input &ex) {
+            error_msg << "Value " << boost::json::serialize(hex_value) << " of '" << key
+                      << "' is not a hexadecimal string";
+            return false;
+        }
+        if constexpr (std::is_same_v<OutputContainer, data_types::Address>) {
+            if (bytes.size() != result.size()) {
+                error_msg << "Wrong size of " << boost::json::serialize(hex_value) << " value of '"
+                          << key << "', expected " << result.size() << " bytes";
+                return false;
+            }
+            std::copy(bytes.begin(), bytes.end(), result.begin());
+        } else {
+            // Store the data of a transaction, could be any size
+            result.clear();
+            for (uint8_t b : bytes) {
+                result.push_back(std::byte(b));
+            }
+        }
+        return true;
+    }
+
+    bool handle_transaction(const std::string &key, const boost::json::value &json_value,
```

Comment:
Overall it seems a bit rough to mix JSON validation logic with actual object generation. Maybe it would be cleaner to separate these two processes so we can develop them separately? E.g. there is a class which is responsible only for input JSON validation. And then it produces some "certainly correct" JSON to actual generator. This may pair well with JSON Schema usage.

Reply:
Done